cmd/kpod/launch.go

```diff
@@ -2,15 +2,568 @@ package main
+// Interface with crio daemon - locking to prevent trampling pod status
+// Launch containers in existing sandboxes
+// Integration tests
+// Man pages
+
 var launchCommand = cli.Command{
 	Name:  "launch",
-	Usage: "launch a pod",
-	Action: func(context *cli.Context) error {
-		return fmt.Errorf("this functionality is not yet implemented")
+	Usage: "launch a pod or insert a container into an existing pod",
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "image",
+			Value: "",
+			Usage: "`image` to launch",
+		},
+		cli.BoolFlag{
+			Name:  "attach",
+			Usage: "`attach` to the primary container once it is created",
+		},
+		cli.StringSliceFlag{
+			Name:  "env",
+			Usage: "specify environment `variable`s to be set inside the primary container, specified as `VARIABLE=VALUE`",
+		},
+		cli.StringFlag{
+			Name:  "labels, l",
+			Value: "",
+			Usage: "specify `label`s to be set on launched pod",
+		},
+		cli.StringFlag{
+			Name:  "limits",
+			Value: "",
+			Usage: "specify resource `limit`s for the primary container",
+		},
+		cli.StringFlag{
+			Name:  "ports",
+			Value: "",
+			Usage: "specify `port`s to be forwarded to the launched pod",
+		},
+		cli.BoolFlag{
+			Name:  "rm",
+			Usage: "`remove` launched container (and pod, if a new pod was created) after it exits",
+		},
+		cli.BoolFlag{
+			Name:  "stdin, i",
+			Usage: "keep `stdin` open on primary container",
+		},
+		cli.BoolFlag{
+			Name:  "tty, t",
+			Usage: "allocate a `TTY` for primary container",
+		},
+		cli.StringSliceFlag{
+			Name:  "mount",
+			Usage: "attach `mount`s on the host to primary container",
+		},
+		cli.StringSliceFlag{
+			Name:  "device",
+			Usage: "make host `device`s available inside the primary container",
+		},
+		cli.StringSliceFlag{
+			Name:  "dns",
+			Usage: "set `DNS server`s for pod",
+		},
+		cli.StringSliceFlag{
+			Name:  "dns-search",
+			Usage: "set `DNS search domain`s for pod",
+		},
+		cli.StringFlag{
+			Name:  "pod",
+			Value: "",
+			Usage: "launch container inside an `existing pod`",
+		},
+		cli.BoolFlag{
+			Name:  "privileged",
+			Usage: "launch a `privileged` pod",
+		},
+		cli.BoolFlag{
+			Name:  "read-only",
+			Usage: "mount root of primary container as `read only`",
+		},
+		cli.BoolFlag{
+			Name:  "host-network",
+			Usage: "don't join a network namespace, and use host's network namespace",
+		},
+		cli.BoolFlag{
+			Name:  "host-ipc",
+			Usage: "don't join an IPC namespace, and use the host's IPC namespace",
+		},
+		cli.BoolFlag{
+			Name:  "host-pid",
+			Usage: "don't join a PID namespace, and use the host's PID namespace",
+		},
+		cli.StringFlag{
+			Name:  "group-add",
+			Value: "",
+			Usage: "comma-separated list of additional `group`s to run as",
+		},
+		cli.StringFlag{
+			Name:  "user",
+			Value: "",
+			Usage: "specify `user` to run primary container as",
+		},
+	},
+	Action: func(ctx *cli.Context) error {
+		if ctx.GlobalBool("debug") {
```

Comment:
Wasn't the original idea that we would refactor out all of the `cri-o` daemon into libraries? I mean, this can be done later but I imagine it's going to be quite ugly to unify the two different bodies of code once we implement everything here.

On the plus side, this code is going to be newer, so maybe we could base the daemon code on this code (after copying it to a library).